src/unicode-utils-posix.cc

```diff
@@ -41,8 +41,11 @@ int UnicodeUtils::bytes_in_characters(const char* string, int characters) {
       break;
 
     bytes += characterLength;
-    if (--characters == 0)
-      break;
+    if (characterLength == 4)
+      characters -= 2;
```

Comment:
This is a key assumption, that a 4-byte character in UTF-8 will always be 2 characters in JavaScript's native UCS-2 encoding. Unsure if this might also sometimes be true for 3-byte UTF-8 characters, or 5 or 6 byte UTF-8 characters. We don't know the rules here so we'll stick to this for now.